File: pt5/1135.connecting-cities-with-minimum-cost.py
from __future__ import annotations

from typing import List


class Solution:
    def minimumCost(self, n: int, connections: List[List[int]]) -> int:
        if n == 1:
            return 0

        parent = list(range(n + 1))
        rank = [0] * (n + 1)

        def find(x: int) -> int:
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x

        def union(a: int, b: int) -> bool:
            root_a = find(a)
            root_b = find(b)
            if root_a == root_b:
                return False
            if rank[root_a] < rank[root_b]:
                root_a, root_b = root_b, root_a
            parent[root_b] = root_a
            if rank[root_a] == rank[root_b]:
                rank[root_a] += 1
            return True

        total = 0
        used = 0
        for a, b, cost in sorted(connections, key=lambda item: item[2]):
            if union(a, b):
                total += cost
                used += 1
                if used == n - 1:
                    return total

        return -1

File: pt5/1135.connecting-cities-with-minimum-cost.java
import java.util.*;

class Solution {
    public int minimumCost(int n, int[][] connections) {
        if (n == 1) {
            return 0;
        }

        Arrays.sort(connections, Comparator.comparingInt(edge -> edge[2]));
        DSU dsu = new DSU(n + 1);
        int total = 0;
        int used = 0;

        for (int[] edge : connections) {
            if (dsu.union(edge[0], edge[1])) {
                total += edge[2];
                used++;
                if (used == n - 1) {
                    return total;
                }
            }
        }

        return -1;
    }

    private static class DSU {
        int[] parent;
        int[] rank;

        DSU(int size) {
            parent = new int[size];
            rank = new int[size];
            for (int i = 0; i < size; i++) {
                parent[i] = i;
            }
        }

        int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        boolean union(int a, int b) {
            int rootA = find(a);
            int rootB = find(b);
            if (rootA == rootB) {
                return false;
            }
            if (rank[rootA] < rank[rootB]) {
                int temp = rootA;
                rootA = rootB;
                rootB = temp;
            }
            parent[rootB] = rootA;
            if (rank[rootA] == rank[rootB]) {
                rank[rootA]++;
            }
            return true;
        }
    }
}

